framework/include/interfaces/FunctorInterface.h

```diff
@@ -0,0 +1,311 @@
+
+  /**
+   * @param face A tuple packing a \p FaceInfo, a \p Limiter, and a boolean denoting whether the
+   * element side of the face is in the upwind direction
+   * @return A limited interpolation to the provided face of the variable or material property. If
+   * the material property is a composite of variables, it's important to note that each variable
+   * will be interpolated individually. If aggregate interpolation is desired, then the user should
+   * make two calls to the \p elem_and_face overload (one for element and one for neighbor) and
+   * then use the global interpolate functions in the \p Moose::FV namespace
+   */
+  virtual T evaluate(const FaceArg & face, unsigned int state) const = 0;
+
+  /**
+   * Evaluate the functor at the current qp point. Unlike the above overloads, there is a caveat to
+   * calling this overload. Any variables involved in the functor evaluation must have their
+   * elemental data properly pre-initialized at the desired quadrature point
+   */
+  virtual T evaluate(const QpArg & qp, unsigned int state) const = 0;
+
+  /**
+   * @param tqp A pair with the first member corresponding to an \p ElementType, either Element,
+   * Neighbor, or Lower corresponding to the three different possible \p MooseVariableData
+   * instances. The second member corresponds to the desired quadrature point
+   * @return The requested element type data indexed at the requested quadrature point. As for the
+   * \p qp overload, there is a caveat: any variables involved in the functor evaluation must have
+   * their requested element data type properly pre-initialized at the desired quadrature point
+   */
+  virtual T evaluate(const std::tuple<Moose::ElementType, unsigned int, SubdomainID> & tqp,
+                     unsigned int state) const = 0;
+
+private:
+  /**
+   * clear cache data
+   */
+  void clearCacheData();
+
+  /// How often to clear the material property cache
+  std::set<ExecFlagType> _clearance_schedule;
+
+  // Data for traditional element-quadrature point property evaluations which are useful for caching
+  // implementation
+
+  /// Current quadrature point element (current cache key)
+  mutable dof_id_type _current_qp_elem = DofObject::invalid_id;
+
+  /// Current quadrature point element data (current cache value)
+  mutable std::vector<std::pair<bool, T>> * _current_qp_elem_data = nullptr;
+
+  /// Cached element quadrature point functor property evaluations. The map key is the element
+  /// id. The map values should have size corresponding to the number of quadrature points on the
+  /// element. The vector elements are pairs. The first member of the pair indicates whether a
+  /// cached value has been computed. The second member of the pair is the (cached) value. If the
+  /// boolean is false, then the value cannot be trusted
+  mutable std::unordered_map<dof_id_type, std::vector<std::pair<bool, T>>> _qp_to_value;
```

Comment:
I don't see how this can work if you're evaluating at multiple different sets of qps on the same element.  This is exactly why we have the elem/side/etc. datastructures in the stateful material properties system.

Reply:
I've added another cache for a new elem-and-side overload. If we get to a point where we key off of hilbert indices, then there will no longer be a need for two caches ... even elem-side evaluations could be cached in an elem-hilbert structure and then this would also allow for doing caching when an element (or an element-and-side pair) get called over and over with *different* quadrature points